torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -300,14 +340,56 @@ void ProcessGroupNCCL::ncclCommWatchdog() {
             // waiting for the operation to complete.
             for (const auto& ncclComm : ncclComms) {
               ncclComm->ncclCommAbort();
+              abortedCommIds.emplace(
+                  buildNcclUniqueIdStr(ncclComm->getNcclId()));
             }
           }
+        }
+      }
+    }
 
-          // Remove communicators from the cache.
-          it = devNCCLCommMap_.erase(it);
+    if (blockingWait_) {
+      // When we abort a communicator on one rank, it is likely that might cause
+      // other ranks to hang indefinitely. As a result, whenever we abort a
+      // communicator, we write its ID to the store. The watchdog on other ranks
+      // then monitor the store, find an aborted communicator ID and abort their
+      // respective communicator as well.
+
+      // Record the aborted communicators locally and in the store.
+      for (const auto& abortedCommId : abortedCommIds) {
+        abortedComms_.emplace(abortedCommId);
+        const auto& storeKey = getNcclAbortedCommStoreKey(abortedCommId);
+        store_->set(storeKey, {});
+        LOG(INFO) << "Watchdog wrote aborted communicator id to store: "
+                  << storeKey;
+      }
 
-        } else {
-          it++;
+      // Check for any communicators in the store and abort them if needed.
+      for (const auto& commId : allCommIds) {
+        if (abortedComms_.find(commId) == abortedComms_.end()) {
+          // Check if we need to abort them if not already aborted (shouldn't
+          // wait more than the watchdog sleep time.).
+          const auto& storeKey = getNcclAbortedCommStoreKey(commId);
+          try {
+            store_->wait(
```

Comment:
What's the default interval you're running this at?

There are 2 caveats with using the store at runtime:
1) If you have a large number of processes all hitting the store at a frequent interval this may cause problems.
2) If you use the TCP store on rank 0, this can timeout/fail if that's the process that went down.

Otherwise, this seems like a decent fix here.

Could you add a couple paragraphs explaining why out of band messaging for aborting NCCL comms is needed? This will help out future readers (including perhaps yourself) on why all of this is needed.